// editorserverapi/lib/audio.js
"use strict";

var ListedEntity = require('./listed_entity');
var FileManager = require('./file_manager');
var fs = require('fs');

/**
* Audio entity
*
* @class Audio
* @constructor
* @param {string} root path
* @param {Express} express application
*/
var Audio = function(_root, _app) {
	ListedEntity.call(this, _root, _app);
};

Audio.prototype = Object.create(ListedEntity.prototype);
Audio.prototype.constructor = Audio;

/**
* Audio key
*
* @property KEY
* @type string
* @default "image"
*/
Audio.KEY = "audio";

/**
* Audio extentions
*
* @property EXTENTIONS
* @type string
* @default []
*/
Audio.EXTENTIONS = [
	".ogg", ".OGG", ".wav", ".WAV", ".mp3", ".MP3"
];

/**
* Get the entity's path
*
* @method getPath
*/
Audio.prototype.getPath = function() {
	return this.root + '/' + Audio.KEY;
};

/**
* Routes the entity
*
* @method routing
*/
Audio.prototype.routing = function() {
	var instance = this;

	this.app.get(this.getPath(), function(req, res) {
		var path = "./app" + req.query.path;
		FileManager.List(path, {recursive: true}, function(_error, _files) {
			res.setHeader('content-type', 'text/plain');
			res.statusCode = 500;

			if (_error) {
				res.write(_error);
			} else {
				res.statusCode = 200;
				res.setHeader('content-type', 'application/json');

				var audios = new Array();
				for (var i = 0; i < _files.length; i++) {
					var file = _files[i];

					try {
						// get image name
						var name = getAudioName(file);

						// add the image
						audios.push({
							name: name,
							path: file
						});
					}	catch(e) {
						console.error(e);
					}
				};

				var data = JSON.stringify({
					audios: audios
				})
				res.write(data);
			}
			res.end();
		});
	});
};

function getAudioName(_file) {
	var name = "Audio";

	var noSlash = _file.replace(/\//g, "-");
	if (noSlash[0] == "-") {
		noSlash = noSlash.substr(1);
	}

	var i = 0;
	var found = false;
	while (i<Audio.EXTENTIONS.length && found == false) {
		var ext = Audio.EXTENTIONS[i];

		if (noSlash.indexOf(ext) > -1) {
			name = noSlash.replace(ext, "");
			found = true;
		}

		i++;
	}
	
	return name;
}

module.exports = Audio;
