// editorserverapi/lib/audio.js
"use strict";

var ListedEntity = require('./listed_entity');
var FileManager = require('./file_manager');
var fs = require('fs');
var path = require('path');

/**
* Audio entity
*
* @class Audio
* @constructor
* @param {string} root path
* @param {Express} express application
*/
var Audio = function(_root, _app) {
	ListedEntity.call(this, _root, _app);
};

Audio.prototype = Object.create(ListedEntity.prototype);
Audio.prototype.constructor = Audio;

/**
* Audio key
*
* @property KEY
* @type string
* @default "image"
*/
Audio.KEY = "audio";

/**
* Allowed audio extentions
*
* @property EXTENTIONS_FILTER
* @type regex
*/
Audio.EXTENTIONS_FILTER = /.(ogg|OGG|wav|WAV|mp3|MP3)/i;

/**
* Get the entity's path
*
* @method getPath
*/
Audio.prototype.getPath = function() {
	return this.root + '/' + Audio.KEY;
};

/**
* Routes the entity
*
* @method routing
*/
Audio.prototype.routing = function() {
	var instance = this;

	this.app.get(this.getPath(), function(req, res) {
		var filepath = "./app" + req.query.path;
		FileManager.List(filepath, {recursive: true}, function(_error, _files) {
			res.setHeader('content-type', 'text/plain');
			res.statusCode = 500;

			if (_error) {
				res.write(_error);
			} else {
				res.statusCode = 200;
				res.setHeader('content-type', 'application/json');

				var audios = new Array();
				for (var i = 0; i < _files.length; i++) {
					var file = _files[i];

					try {
						// get file extension
						var extension = path.extname(file);

						// if the extenstion is allowed
						if (Audio.EXTENTIONS_FILTER.test(extension)) {
							var audioNameWithDashes = file.replace(/\//g, "-");
							if (audioNameWithDashes[0] == "-") {
								audioNameWithDashes = audioNameWithDashes.substr(1);
							}

							// add the audio
							audios.push({
								name: audioNameWithDashes,
								path: file
							});
						}
					}	catch(e) {
						console.error(e);
					}
				};

				var data = JSON.stringify({
					audios: audios
				})
				res.write(data);
			}
			res.end();
		});
	});
};

module.exports = Audio;
