// editorserverapi/lib/image.js
"use strict";

var ListedEntity = require('./listed_entity');
var FileManager = require('./file_manager');
var fs = require('fs');

/**
* Image entity
*
* @class Image
* @constructor
* @param {string} root path
* @param {Express} express application
*/
var Image = function(_root, _app) {
	ListedEntity.call(this, _root, _app);
};

Image.prototype = Object.create(ListedEntity.prototype);
Image.prototype.constructor = Image;

/**
* Image key
*
* @property KEY
* @type string
* @default "image"
*/
Image.KEY = "image";

/**
* Image extentions
*
* @property EXTENTIONS
* @type string
* @default []
*/
Image.EXTENTIONS = [
	".png", ".PNG", ".jpg", ".JPG", ".jpeg", ".JPEG", ".gif", ".GIF"
];

/**
* Get the entity's path
*
* @method getPath
*/
Image.prototype.getPath = function() {
	return this.root + '/' + Image.KEY;
};

/**
* Routes the entity
*
* @method routing
*/
Image.prototype.routing = function() {
	var instance = this;

	this.app.get(this.getPath(), function(req, res) {
		var path = "./app" + req.query.path;
		FileManager.List(path, {recursive: true}, function(_error, _files) {
			res.setHeader('content-type', 'text/plain');
			res.statusCode = 500;

			if (_error) {
				res.write(_error);
			} else {
				res.statusCode = 200;
				res.setHeader('content-type', 'application/json');

				var images = new Array();
				for (var i = 0; i < _files.length; i++) {
					var file = _files[i];

					try {
						// get image name
						var name = getImageName(file);

						// add the image
						images.push({
							name: name,
							path: file
						});
					}	catch(e) {
						console.error(e);
					}
				};

				var data = JSON.stringify({
					images: images
				})
				res.write(data);
			}
			res.end();
		});
	});
};

function getImageName(_file) {
	var name = "Image";

	var noSlash = _file.replace(/\//g, "-");
	if (noSlash[0] == "-") {
		noSlash = noSlash.substr(1);
	}

	var i = 0;
	var found = false;
	while (i<Image.EXTENTIONS.length && found == false) {
		var ext = Image.EXTENTIONS[i];

		if (noSlash.indexOf(ext) > -1) {
			name = noSlash.replace(ext, "");
			found = true;
		}

		i++;
	}
	
	return name;
}

module.exports = Image;
