// editorserverapi/lib/index.js
"use strict";

var Project = require('./project');
var Level = require('./level');
var Prefab = require('./prefab');
var Layers = require('./layers');
var Inputs = require('./inputs');
var Behaviour = require('./behaviour');
var Image = require('./image');
var Audio = require('./audio');
var BitmapFont = require('./bitmapfont');
var Atlas = require('./atlas');

/**
* Server API for editor
*
* @class EditorServerAPI
* @constructor
* @param {Express} express application
*/
var EditorServerAPI = function(_app) {
	this.app = _app;

	this.root = '/' + EditorServerAPI.KEY + '/' + EditorServerAPI.VERSION;
	this.entities = new Array();

	this.loadEntities();
};

EditorServerAPI.KEY = "editorserverapi";
EditorServerAPI.VERSION = "v0";

EditorServerAPI.prototype.loadEntities = function() {
	this.entities.push(new Project(this.root, this.app));
	this.entities.push(new Level(this.root, this.app));
	this.entities.push(new Prefab(this.root, this.app));
	this.entities.push(new Layers(this.root, this.app));
	this.entities.push(new Inputs(this.root, this.app));
	this.entities.push(new Behaviour(this.root, this.app));
	this.entities.push(new Image(this.root, this.app));
	this.entities.push(new Audio(this.root, this.app));
	this.entities.push(new BitmapFont(this.root, this.app));
	this.entities.push(new Atlas(this.root, this.app));
};

/**
* Routes the entities
*
* @method routing
*/
EditorServerAPI.prototype.routing = function() {
	if (this.entities) {
		for (var i = 0; i < this.entities.length; i++) {
			var ent = this.entities[i];
			if (typeof ent.routing === "function") {
				ent.routing();
			}
		};
	}
};


module.exports = EditorServerAPI;
