// editorserverapi/lib/prefab.js
"use strict";

var ListedEntity = require('./listed_entity');
var FileManager = require('./file_manager');
var fs = require('fs');
var path = require('path');

/**
* Prefab entity
*
* @class Prefab
* @constructor
* @param {string} root path
* @param {Express} express application
*/
var Prefab = function(_root, _app) {
	ListedEntity.call(this, _root, _app);
};

Prefab.prototype = Object.create(ListedEntity.prototype);
Prefab.prototype.constructor = Prefab;

/**
* Prefab key
*
* @property KEY
* @type string
* @default "prefab"
*/
Prefab.KEY = "prefab";

/**
* Allowed prefab extentions
*
* @property EXTENTIONS_FILTER
* @type regex
*/
Prefab.EXTENTIONS_FILTER = /.(json)/i;

/**
* Get the entity's path
*
* @method getPath
*/
Prefab.prototype.getPath = function() {
	return this.root + '/' + Prefab.KEY;
};

/**
* Routes the entity
*
* @method routing
*/
Prefab.prototype.routing = function() {
	var instance = this;

	var instance = this;

	this.app.get(this.getPath(), function(req, res) {
		var filepath = "./app" + req.query.path;
		FileManager.List(filepath, {recursive: true}, function(_error, _files) {
			res.setHeader('content-type', 'text/plain');
			res.statusCode = 500;

			if (_error) {
				res.write(_error);
			} else {
				res.statusCode = 200;
				res.setHeader('content-type', 'application/json');

				var prefabs = new Array();
				for (var i = 0; i < _files.length; i++) {
					var file = _files[i];

					try {
						// get file extension
						var extension = path.extname(file);

						// if the extenstion is allowed
						if (Prefab.EXTENTIONS_FILTER.test(extension)) {
							// add the audio
							prefabs.push({
								name: path.basename(file, extension),
								path: file
							});
						}
					}	catch(e) {
						console.error(e);
					}
				};

				var data = JSON.stringify({
					prefabs: prefabs
				})
				res.write(data);
			}
			res.end();
		});
	});

	this.app.get(this.getPath() + "/:name", function(req, res) {
		var path = "./app" + req.query.path;
		var name = req.params.name + ".json";
		FileManager.Read2(path, name, function(error, data) {
			if (error) {
				res.setHeader('content-type', 'text/plain');
				res.statusCode = 500;
				res.write(error);
			} else {
				res.statusCode = 200;
				res.setHeader('content-type', 'application/json');
				res.write(data);
			}
			res.end();
		});
	});

	this.app.post(this.getPath(), function(req, res) {
		FileManager.Create(req, res, function(_req, _res) {
			_res.end();
		});
	});
};

module.exports = Prefab;
